[CORE-13211] iceberg/conversion: Map PB enum values to string, not int

wdberkeley opened the pull request:
Map protobuf enum values to their string name instead of int value. For example, the enum

    enum state {
        Active = 0;
        Inactive = 1;
    }

previously would translate `Active = 0` to `0` and `Inactive = 1` to `1`, but now will translate `Active = 0` to `"ACTIVE"` and `Inactive = 1` to `"INACTIVE"` (note the capitalization, which comes from the PB descriptor).

This is a breaking change. The type of columns translated from enums will change from signed integer to string.

Values that are not valid for the enum are translated to the default value's name. For example, a value of 100 for `state`
will be translated into `"ACTIVE"`.

Fixes CORE-13211

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* Iceberg now translates protobuf enum values to their name, instead of integer value. This is a backwards-incompatible change. Additionally, it is now an error to provide an enum value outside the range of the enum definition, as it has no name it can be converted into.

wdberkeley committed:
iceberg/conversion: Map PB enum values to string, not int

Map protobuf enum values to their string name instead of int value. For
example, the enum

    enum state {
        Active = 0;
        Inactive = 1;
    }

previously would translate `Active = 0` to `0` and `Inactive = 1` to `1`,
but now will translate `Active = 0` to `"ACTIVE"` and `Inactive = 1` to
`"INACTIVE"` (note the capitalization, which comes from the PB
descriptor).

This is a breaking change. The type of columns translated from enums
will change from signed integer to string.

This also introduces a new error condition: previously, any integer
would be accepted as an enum value and translated to itself. Now, lookup
of the enum name will fail if the value isn't in the range of the enum.
This is a translation error because it means the value is invalid, at
least for the current schema. We could also handle it by using the enum
default, or by stringifying the value (`99` -> `"99"`).

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--8691aeb9-9b32-4f76-b98d-a9acda13b951-->
<!--TEST RESULTS-->
<!--93553837-f991-4a01-810c-1f53ffc6a1c5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73786">build#73786</a></summary>


| test_class                    | test_method            | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                       |
|:------------------------------|:-----------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73786#0199c4e5-168c-4d48-b908-4440931d19f2 | FLAKY         | 17/21    | upstream reliability is '95.79500657030223'. current run reliability is '80.95238095238095'. drift is 14.84263 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74370">build#74370</a></summary>


| test_class                          | test_method              | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                             | test_history                                                                                                                                                           |
|:------------------------------------|:-------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConfigurationUpdateTest             | test_two_nodes_update    | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/74370#0199f08f-bd1b-442b-95ef-db1e5e3bc4f9 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConfigurationUpdateTest&test_method=test_two_nodes_update            |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74370#0199f092-11e3-4398-b722-4aa4826ee7d1 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| WriteCachingFailureInjectionE2ETest | test_crash_all           | {"use_transactions": false}                        | integration | https://buildkite.com/redpanda/redpanda/builds/74370#0199f092-11e7-4545-a427-b0953b5abc2d | FLAKY         | 17/21    | upstream reliability is '81.811377245509'. current run reliability is '80.95238095238095'. drift is 0.859 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all       |
</details>
<!--/body-->

andrwng reviewed (commented):
This code looks good, but need some input from Matt on the ticket about how far we should push on making this safe across upgrades.

andrwng commented:
>Additionally, it is now an error to provide an enum value outside the range of the enum definition, as it has no name it can be converted into.

What happens in this case? Mind adding a test case for it?

EDIT: oh this is probably the new value_conversion_exception scenario, still a test would be great!

rockwotj commented:
> > Additionally, it is now an error to provide an enum value outside the range of the enum definition, as it has no name it can be converted into.
> 
> What happens in this case? Mind adding a test case for it?
> 
> EDIT: oh this is probably the new value_conversion_exception scenario, still a test would be great!

Relevant reading: https://protobuf.dev/programming-guides/enum/ and the reason I didn't use string values. I think we want to treat enums as *closed* always, that feels better than DLQ and we don't have the option to treat them as open.

rockwotj commented:
Also, we need to hold off on this, we need to check on this breaking change that it's OK to make.

wdberkeley committed:
iceberg/conversion: Convert invalid enums to default

This alters the error handling of the previous commit so that unknown
enum values are converted to the default enum value's name, instead of
being a translation error. A new test case is included.

wdberkeley commented:
Force push to resolve conflict in tests.
